stl/inc/format

```diff
@@ -389,26 +555,256 @@ public:
+// in order to fetch the width from the arguments.
+template <class _ParseContext, class _Context>
+class _Specs_handler : public _Specs_setter<typename _Context::char_type> {
+public:
+    using _CharT = typename _Context::char_type;
+
+    constexpr _Specs_handler(_Basic_format_specs<_CharT>& _Specs, _ParseContext& _Parse_ctx, _Context& _Ctx)
+        : _Specs_setter<_CharT>(_Specs), _Parse_ctx(_Parse_ctx), _Ctx(_Ctx) {}
+
+    template <class _Id>
+    constexpr void _On_dynamic_width(_Id _Arg_id) {
+        this->_Specs._Width = _Get_dynamic_specs<_Width_checker>(_Get_arg(_Arg_id));
+    }
+
+    template <class _Id>
+    constexpr void _On_dynamic_precision(_Id _Arg_id) {
+        this->_Specs._Precision = _Get_dynamic_specs<_Precision_checker>(_Get_arg(_Arg_id));
+    }
+
+private:
+    _ParseContext& _Parse_ctx;
+    _Context& _Ctx;
+
+    constexpr basic_format_arg<_Context> _Get_arg(_Auto_id_tag) {
+        return _STD _Get_arg(_Ctx, _Parse_ctx.next_arg_id());
+    }
+
+    constexpr basic_format_arg<_Context> _Get_arg(size_t _Arg_id) {
+        _Parse_ctx.check_arg_id(_Arg_id);
+        return _STD _Get_arg(_Ctx, _Arg_id);
+    }
 };
+class _Numeric_specs_checker {
+    _Type _Arg_type = _Type::_None;
+
+public:
+    constexpr _Numeric_specs_checker(_Type _Arg_type) : _Arg_type(_Arg_type) {}
+
+    constexpr void _Require_numeric_argument() {
+        if (!_Is_numeric_fmt_type(_Arg_type)) {
+            throw format_error("Format specifier requires numeric argument.");
+        }
+    }
+
+    constexpr void _Check_sign() {
+        _Require_numeric_argument();
+        if (_Is_integral_fmt_type(_Arg_type) && _Arg_type != _Type::_Int32 && _Arg_type != _Type::_Int64
+            && _Arg_type != _Type::_Char) {
+            throw format_error("Format specifier requires signed argument.");
+        }
+    }
+
+    constexpr void _Check_precision() {
+        if (_Is_integral_fmt_type(_Arg_type) || _Arg_type == _Type::_Pointer) {
+            throw format_error("Precision not allowed for this argument type.");
+        }
+    }
+};
+
+// Uses _Numeric_specs_checker to check that the type of the argument printed by
+// a replacement field with format specs actually satisfies the requirements for
+// that format spec. If the requirements are met then calls the base class
+// handler method.
+template <class _Handler>
+class _Specs_checker : public _Handler {
+    _Numeric_specs_checker _Numeric_checker;
+
+public:
+    constexpr _Specs_checker(const _Handler& _Handler_inst, _Type _Arg_type)
+        : _Handler(_Handler_inst), _Numeric_checker(_Arg_type) {}
+
+    // _On_align has no checking, since we don't implement numeric alignments.
+    constexpr void _On_sign(_Sign _Sgn) {
+        _Numeric_checker._Check_sign();
+        _Handler::_On_sign(_Sgn);
+    }
+
+    constexpr void _On_hash() {
+        _Numeric_checker._Require_numeric_argument();

```

Comment:
no, it doesn't. That has to be checked later since you can have a hash with a bool or char if you've passed a numeric "type" later on. I'll add a comment